norddrop/src/device.rs

```diff
@@ -522,22 +522,29 @@ impl NordDropFFI {
     }
 }
 
-fn crate_key_context(
-    logger: slog::Logger,
-    privkey: SecretKey,
-    pubkey_cb: impl Fn(IpAddr) -> Option<PublicKey> + Send + 'static,
-) -> auth::Context {
-    let pubkey_cb = std::sync::Mutex::new(pubkey_cb);
-    let public = move |ip: IpAddr| {
-        let guard = pubkey_cb.lock().expect("Failed to lock pubkey callback");
-        let key = guard(ip)?;
-        drop(guard);
+fn crate_key_context(logger: slog::Logger, key_store: Arc<Box<dyn KeyStore>>) -> auth::Context {
+    let priv_keystore = Arc::clone(&key_store);
+
+    let priv_logger = logger.clone();
+    let privkey = move || {
+        let key_store = Arc::clone(&priv_keystore);
+
+        let privkey: [u8; PUBLIC_KEY_LENGTH] = key_store.privkey().try_into().ok()?;
+        debug!(priv_logger, "Retrieved private key: {:?}", privkey);
+        Some(SecretKey::from(privkey))
+    };
+
+    let pub_keystore = Arc::clone(&key_store);
+    let pubkey = move |ip: IpAddr| {
+        let key_store = Arc::clone(&pub_keystore);
 
-        debug!(logger, "Public key for {ip:?}: {key:?}");
-        Some(key)
+        let pubkey = key_store.on_pubkey(ip.to_string())?;
+        let pubkey: [u8; PUBLIC_KEY_LENGTH] = pubkey.try_into().ok()?;
+        debug!(logger, "Retrieved public key for: {} key: {:?}", ip, pubkey);
+        Some(PublicKey::from(pubkey))
     };
 
-    auth::Context::new(privkey, public)
+    auth::Context::new(privkey, pubkey)
```

Comment:
Is this fine in general? the previous version used a mutex here, is it okay to remove it here?

Reply:
I was thinking about this mutex before removal quite hard and I can only see one reason for it being there - to serialize libdrop's callbacks, meaning if two would happen simultaneously, they would go one after another. I don't see any reason for this being a problem at all to be honest.

As per mutex - it should be implemented in KeyStore implementation on the app side that actually controls the keys, it should then lock the mutex on the callback and when updating it. Guarding with mutex in libdrop just serializes the callback invocations for no purpose IMO. @matszczygiel might have an opinion on this

Reply:
oh didn't notice, he's on vacation.

Reply:
Hey, IMHO Mutex should be there since we cannot trust apps to provide thread-safe callback

Reply:
Imagine app fetching the key from hash map in the callback and at the same time updating the hash map on some other thread with new keys or more peers

Reply:
But the problem is that it doesn't prevent the race, we do not lock the mutex when there's an update of the value(as we don't modify ourselves), only when reading, thus we basically just prevent libdrop from reading in parallel but don't guard against the modifications unless I miss something?

Reply:
Yes, we cannot fully protect from apps doing data races, but at least we can defend as much as we can like the callback is modifying something (on the app side) without mutex protection. We cannot guarantee this doesn't happen

Reply:
Mutex only makes sense when both readers and writers are aware of it and block on locking, thus if the app is not aware of our internal mutex, I don't see the benefit we get. I would speak differently if libdrop would also modify the keys in the callback, in that case our mutex at least would guard to some degree IMO

Reply:
The example where the app is modifying a data structure in the callback, and the mutex is protecting us from data race
```cpp
class Callback: KeyStore {
    private:
        Cache _cache;
        DB _db;
    public:
        vector<char> on_pubkey(const char* peer) override {
            if _cache.contains(peer) {
                return _cache.key(peer);
            }

            auto key = _db.fetch(peer);
            _cache.insert(peer, key); // <-- this is writing to some memory structure
            return key;
        }
}
```

Reply:
Restored the Mutex on Public-Key retrieval. Skipped it on the Private-Key retrieval